src/Core/Services/EnvironmentService.cs

```diff
@@ -96,15 +110,57 @@ public async Task SetUrlsFromStorageAsync()
 
         }
 
+        private async Task<Region?> MigrateToRegionsAsync(EnvironmentUrlData urls)
+        {
+            if (urls.Base == Region.US.BaseUrl())
+            {
+                await _stateService.UpdateActiveUserEnvironmentAsync(Region.US, Region.US.GetUrls());
+                return Region.US;
+            }
+            if (urls.Base == Region.EU.BaseUrl())
+            {
+                await _stateService.UpdateActiveUserEnvironmentAsync(Region.EU, Region.EU.GetUrls());
+                return Region.EU;
+            }
+            await _stateService.UpdateActiveUserEnvironmentAsync(Region.SelfHosted, urls);
+            return Region.SelfHosted;
+        }
+
         public async Task<EnvironmentUrlData> SetUrlsAsync(EnvironmentUrlData urls)
         {
-            urls.Base = FormatUrl(urls.Base);
-            urls.WebVault = FormatUrl(urls.WebVault);
-            urls.Api = FormatUrl(urls.Api);
-            urls.Identity = FormatUrl(urls.Identity);
-            urls.Icons = FormatUrl(urls.Icons);
-            urls.Notifications = FormatUrl(urls.Notifications);
-            urls.Events = FormatUrl(urls.Events);
+            // format urls for Base comparison
+            urls = FormatUrls(urls);
+            if (urls.Base == Region.EU.BaseUrl())
+            {
+                return await SetRegionAsync(Region.EU);
+            }
+            if (urls.Base == Region.US.BaseUrl())
+            {
+                return await SetRegionAsync(Region.US);
+            }
+            return await SetRegionAsync(Region.SelfHosted, urls);
+        }
+
+        public async Task<EnvironmentUrlData> SetRegionAsync(Region region, EnvironmentUrlData selfHostedUrls = null)
+        {
+            EnvironmentUrlData urls;
+            await _stateService.SetPreAuthRegionAsync(region);
+            SelectedRegion = region;
+
+            if (region == Region.SelfHosted)
+            {
+                // If user saves a self-hosted region with empty fields, default to US
+                if (selfHostedUrls.IsEmpty)
+                {
+                    return await SetRegionAsync(Region.US);
+                }
+                urls = FormatUrls(selfHostedUrls);
+            }
+            else
+            {
+                urls = region.GetUrls();
+            }
+
```

Comment:
🤔 To avoid racing conditions, what do you think about changing the order this is done given that if the self hosted Uris are empty then US will be set afterwards, but in the meantime PreAuth would have been set to SelfHost which would be wrong, so I imagine something like this:
```suggestion
            EnvironmentUrlData urls;
            
            if (region == Region.SelfHosted)
            {
                // If user saves a self-hosted region with empty fields, default to US
                if (selfHostedUrls.IsEmpty)
                {
                    return await SetRegionAsync(Region.US);
                }
                urls = FormatUrls(selfHostedUrls);
            }
            else
            {
                urls = region.GetUrls();
            }
            
            await _stateService.SetPreAuthRegionAsync(region);
            SelectedRegion = region;

```
Addionally, this would be theoretically faster in the case self hosted urls are empty because we are awaiting only one time setting pre auth region.